bittide/src/Bittide/Calculator.hs

```diff
@@ -0,0 +1,571 @@
+chainConfigurationWorker _ fpgaConfig ugnParts writeOffset =
+  leToPlus @1 @nNodes
+    $ leToPlus @1 @chainLength
+    $ leToPlusKN @chainLength @nNodes
+    $ let
+        start = (0, writeOffset)
+        ((_, startReadAt), configs) = mapAccumL mkLink start (takeI indicesI)
+        (firstConfig, restConfigs) = uncons configs
+        readForN = cyclesPerWrite * natToNum @chainLength
+        (readMetacycle, readOffset) = quotRem startReadAt metacycleLength
+       in
+        firstConfig
+          { readMetacycle = readMetacycle
+          , readOffset = readOffset
+          , readForN = readForN
+          }
+          :> restConfigs
+ where
+  cyclesPerWrite :: Integer
+  cyclesPerWrite = natToNum @cyclesPerWrite
+  groupCycles :: Integer
+  groupCycles = natToNum @(GroupCycles nNodes cyclesPerWrite)
+  windowCycles :: Integer
+  windowCycles = natToNum @(WindowCycles nNodes cyclesPerWrite)
+  metacycleLength :: Integer
+  metacycleLength = natToNum @(MetacycleLength nNodes cyclesPerWrite padding reps)
+
+  counterMap :: Vec nNodes (Map (Index nNodes) Integer)
+  counterMap = toCounterMap (toFpgaIndexed fpgaConfig ugnParts)
+
+  mkLink ::
+    (Integer, Integer) ->
+    Index nNodes ->
+    ((Integer, Integer), MetaPeConfig Integer Integer Integer)
+  mkLink (readForN, startReadAt) node =
+    ((writeForN, nextStartReadAt), peConfig)
+   where
+    writeForN = readForN + cyclesPerWrite
+    (rM, rO) = quotRem startReadAt metacycleLength
+
+    -- Writes have to happen in the _next_ metacycle, so adjust for that.
+    startWriteAtClosestTo = roundUpMetacycle $ startReadAt + readForN
+    k = findK node nextNode startWriteAtClosestTo
+    startWriteAt = startWriteAtClosestTo + k
+    (wM, wO) = quotRem startWriteAt metacycleLength
+    peConfig =
+      MetaPeConfig
+        { writeMetacycle = wM
+        , writeOffset = wO
+        , readMetacycle = rM
+        , readOffset = rO
+        , ..
+        }
+
+    nextStartReadAt = mapCycle startWriteAt node nextNode
+    nextNode = satSucc SatWrap node
+
+  roundUpMetacycle :: Integer -> Integer
+  roundUpMetacycle cyc = metacycleLength * (cyc `divRU` metacycleLength)
+
+  mapCycle :: Integer -> Index nNodes -> Index nNodes -> Integer
+  mapCycle srcCycle src dst = srcCycle + counterMap !! src Map.! dst
+
+  findK :: Index nNodes -> Index nNodes -> Integer -> Integer
+  findK src dst srcCycle
+    | offsetInDstMetacycle <= linkSelectOffset0 = linkSelectOffset0 - offsetInDstMetacycle
+    | offsetInDstMetacycle <= linkSelectOffset1 = linkSelectOffset1 - offsetInDstMetacycle
+    | otherwise = linkSelectOffset0 + metacycleLength - offsetInDstMetacycle
+   where
+    -- For a graphical representation of what we're doing here, check out:
+    -- bittide-instances/imgs/ugn_calculator.drawio. For a general introduction
+    -- see slides "determining read/write cycles" in
+    -- https://docs.google.com/presentation/d/1AGbAJQ1zhTPtrekKnQcthd0TUPyQs-zowQpV1ux4k-Y
```

Comment:
Could you update the drawio diagram? Could you copy the presentation and create an updated one?